### ISSUE TITLE
Bug with combination of contextmanager, partial, and ParamSpec

### ISSUE THREAD

**Issue by dmontagu** [microsoft/pyright#10011]
**Describe the bug**
If I have a generic dataclass with a `ParamSpec` in its generic parameters, and it has a method decorated with `@contextmanager` or `@asynccontextmanager` that accepts the ParamSpec's args/kwargs, and I use a `functools.partial` as the input (which is used to infer the `ParamSpec`), there seems to be a bug where pyright thinks I need to pass a value for `self` on the method. Minimal reproduction below.

Please let me know if I'm misunderstanding something, but as far as I can tell the issue seems to be coming from the combination of all three of these things (ParamSpec, contextmanager, and partial). **If I remove any one of these, the spurious type error goes away.**

**Code or Screenshots**
The following examples execute without errors and as far as I can tell should not produce type errors. I've included the pyright errors and printed outputs in in-line comments.

Sync example:
```python
import dataclasses
from contextlib import contextmanager
from functools import partial
from typing import ParamSpec, Generic, Callable, Iterator

P = ParamSpec('P')


@dataclasses.dataclass
class MyCaller(Generic[P]):
    my_callable: Callable[P, None]

    @contextmanager
    def my_context_manager_method(self, *args: P.args, **kwargs: P.kwargs) -> Iterator[None]:
        self.my_callable(*args, **kwargs)
        yield


def my_callable(x: int, y: int) -> None:
    print(f'{x=} {y=}')


def main():
    task = partial(my_callable, y=2)
    with MyCaller(task).my_context_manager_method(x=1):  # error: Argument missing for parameter "self" (reportCallIssue)
        pass

main()
#> x=1 y=2
```

Async example:
```python
import asyncio 
import dataclasses
from contextlib import asynccontextmanager
from functools import partial
from typing import ParamSpec, Generic, Callable, AsyncIterator, Awaitable

P = ParamSpec('P')


@dataclasses.dataclass
class MyCaller(Generic[P]):
    my_callable: Callable[P, Awaitable[None]]

    @asynccontextmanager
    async def my_context_manager_method(self, *args: P.args, **kwargs: P.kwargs) -> AsyncIterator[None]:
        await self.my_callable(*args, **kwargs)
        yield


async def my_callable(x: int, y: int) -> None:
    print(f'{x=} {y=}')


async def main():
    task = partial(my_callable, y=2)
    async with MyCaller(task).my_context_manager_method(x=1):  # error: Argument missing for parameter "self" (reportCallIssue)
        pass


asyncio.run(main())
#> x=1 y=2
```

**VS Code extension or command-line**
Command line. I get this behavior on both v1.1.389 and v1.1.395.

----

Pyright is awesome, thank you so much!

**Comment by dmontagu** [comment]
**Reductions of the above that _do NOT_ produce type errors:**

No `@contextmanager`:
```python
import dataclasses
from functools import partial
from typing import ParamSpec, Generic, Callable, Iterator

P = ParamSpec('P')


@dataclasses.dataclass
class MyCaller(Generic[P]):
    my_callable: Callable[P, None]

    def my_context_manager_method(self, *args: P.args, **kwargs: P.kwargs) -> Iterator[None]:
        self.my_callable(*args, **kwargs)
        yield


def my_callable(x: int, y: int) -> None:
    print(f'{x=} {y=}')


def main():
    task = partial(my_callable, y=2)
    MyCaller(task).my_context_manager_method(x=1)

main()
#> x=1 y=2
```

No `ParamSpec`:
```python
import dataclasses
from contextlib import contextmanager
from functools import partial
from typing import Callable, Iterator, Any


@dataclasses.dataclass
class MyCaller:
    my_callable: Callable[..., None]

    @contextmanager
    def my_context_manager_method(self, *args: Any, **kwargs: Any) -> Iterator[None]:
        self.my_callable(*args, **kwargs)
        yield


def my_callable(x: int, y: int) -> None:
    print(f'{x=} {y=}')


def main():
    task = partial(my_callable, y=2)
    with MyCaller(task).my_context_manager_method(x=1):
        pass


main()
# > x=1 y=2
```

No `partial`:
```python
import dataclasses
from contextlib import contextmanager
from functools import partial
from typing import ParamSpec, Generic, Callable, Iterator

P = ParamSpec('P')


@dataclasses.dataclass
class MyCaller(Generic[P]):
    my_callable: Callable[P, None]

    @contextmanager
    def my_context_manager_method(self, *args: P.args, **kwargs: P.kwargs) -> Iterator[None]:
        self.my_callable(*args, **kwargs)
        yield


def my_callable(x: int, y: int) -> None:
    print(f'{x=} {y=}')


def main():
    with MyCaller(my_callable).my_context_manager_method(x=1, y=2):
        pass

main()
#> x=1 y=2
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.398.
